### ISSUE TITLE
NCC in CCM is not correctly computed

### ISSUE THREAD

**Issue by victornoel** [yegor256/jpeek#449]
While investigating #447, I noticed a potential error in the xsl file, but I'm not 100% sure it is an error or not because I'm not very good with xsl.

Actually this error has no impact to solve #447 because the number it find is the correct one even though the calculation is wrong. There was another error somewhere else.

Basically, NCC is meant to be the number of connected components (of the graph where methods are the nodes and are connected according to some criteria). See for example https://www.geeksforgeeks.org/connected-components-in-an-undirected-graph/ to understand how to count them. It is not the number of nodes.

NCC is the xsl is defined like this, and it seems to count the number of nodes:
```xsl
<xsl:variable name="ncc" select="count(distinct-values($edges/edge/method/text()))"/>
```

Furthermore it counts them incorrectly because the name of the method is used and some methods have the same name in some of the test classes... ~~I will create another issue for that problem because it pertains to other metrics I think.~~ actually it doesn't seem to be the case.

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) please, pay attention to this issue

**Comment by paulodamaso** [comment]
@victornoel Could you please fix the title? It is just related to NCC right?

**Comment by victornoel** [comment]
@paulodamaso yes sorry, it was `in` not `is` :)

**Comment by paulodamaso** [comment]
@victornoel Thank you!

**Comment by paulodamaso** [comment]
@0crat in

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616475800))

@paulodamaso Election failed: performer was not elected (see footprint for details)

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616475800))

@paulodamaso Election failed: performer was not elected (see footprint for details)

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616475800))

@paulodamaso Election failed: performer was not elected (see footprint for details)

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616475800))

@paulodamaso Election failed: performer was not elected (see footprint for details)

**Comment by 0crat** [comment]
> @0crat in ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616475800))

@paulodamaso Election failed: performer was not elected (see footprint for details)

**Comment by 0crat** [comment]
The job [#449](https://github.com/yegor256/jpeek/issues/449) assigned to @hdouss[/z](https://www.0crat.com/u/hdouss), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/8bfb092e-0c96-48c6-98e2-2c5d2f2eb3bb); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by HDouss** [comment]
@victornoel it is actually calculating the number of nodes having an edge. And yes, it is completely erroneous.

**Comment by HDouss** [comment]
@0crat refuse will have to wait too much for #437 and its puzzles..

**Comment by 0crat** [comment]
> @0crat refuse will have to wait too much for #437 and its puzzles.. ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616771321))

@HDouss The user @hdouss[/z](https://www.0crat.com/u/hdouss) resigned from [#449](https://github.com/yegor256/jpeek/issues/449), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse will have to wait too much for #437 and its puzzles.. ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-616771321))

@HDouss Job refused in 0 hours - no penalty, see [§6](http://www.zerocracy.com/policy.html#6)

**Comment by 0crat** [comment]
The job [#449](https://github.com/yegor256/jpeek/issues/449) assigned to @dfrau32[/z](https://www.0crat.com/u/dfrau32), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/b83e2b1f-b641-47d0-8dbc-094a08e7d3f7); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by 0crat** [comment]
@dfrau32[/z](https://www.0crat.com/u/dfrau32) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
The user @dfrau32[/z](https://www.0crat.com/u/dfrau32) resigned from [#449](https://github.com/yegor256/jpeek/issues/449), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
The job [#449](https://github.com/yegor256/jpeek/issues/449) assigned to @vzurauskas[/z](https://www.0crat.com/u/vzurauskas), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/9c3c878e-5b88-445c-9738-b6a80b63f288); the budget is 30 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§8](http://www.zerocracy.com/policy.html#8) and [§9](http://www.zerocracy.com/policy.html#9); if the task is not clear, read [this](http://www.yegor256.com/2015/02/16/it-is-not-a-school.html) and [this](http://www.yegor256.com/2015/01/15/how-to-cut-corners.html); there will be a monetary reward for this job

**Comment by vzurauskas** [comment]
@paulodamaso If I understand correctly, the `Graph` interface was created in order to do calculations like the one required here because graphs are better handled in Java than XSL, correct? If so, would it make sense to rewrite CCM metric in Java instead of fixing this calculation in the XSL?

**Comment by vzurauskas** [comment]
@0crat wait for clarification.

**Comment by paulodamaso** [comment]
@vzurauskas Yes, it makes sense, let's use java to make these calculations that need graphs

**Comment by vzurauskas** [comment]
@victornoel https://github.com/yegor256/jpeek/pull/477 was merged and `todo`s created, can this be closed now?

**Comment by victornoel** [comment]
@vzurauskas thx!

**Comment by 0crat** [comment]
@vzurauskas[/z](https://www.0crat.com/u/vzurauskas) resigned from [#449](https://github.com/yegor256/jpeek/issues/449), since the job is not in scope anymore

**Comment by vzurauskas** [comment]
@paulodamaso Zerocrat resigned me, but should have asked for quality review and paid me after that.

**Comment by paulodamaso** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-641209419))

@paulodamaso This is what I know about this job in C7JGJ00DP, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#449](https://github.com/yegor256/jpeek/issues/449) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C7JGJ00DP?q=%7Bjob%3A%27gh%3Ayegor256%2Fjpeek%23449%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by paulodamaso** [comment]
@0crat in

**Comment by vzurauskas** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-642607097))

@vzurauskas This is what I know about this job in C7JGJ00DP, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#449](https://github.com/yegor256/jpeek/issues/449) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C7JGJ00DP?q=%7Bjob%3A%27gh%3Ayegor256%2Fjpeek%23449%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by vzurauskas** [comment]
@paulodamaso The ticket is still stuck.

**Comment by vzurauskas** [comment]
@paulodamaso ping

**Comment by 0pdd** [comment]
@victornoel 2 puzzles [#481](https://github.com/yegor256/jpeek/issues/481), [#482](https://github.com/yegor256/jpeek/issues/482) are still not solved.

**Comment by vzurauskas** [comment]
@paulodamaso ping

**Comment by paulodamaso** [comment]
@0crat quality good

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/yegor256/jpeek/issues/449#issuecomment-664031864))

@paulodamaso There is no quality review for [#449](https://github.com/yegor256/jpeek/issues/449), no performer

**Comment by 0pdd** [comment]
@victornoel 2 puzzles [#511](https://github.com/cqfn/jpeek/issues/511), [#512](https://github.com/cqfn/jpeek/issues/512) are still not solved.

**Comment by 0pdd** [comment]
@victornoel the puzzle [#511](https://github.com/cqfn/jpeek/issues/511) is still not solved; solved: [#512](https://github.com/cqfn/jpeek/issues/512).
